### PR TITLE
Round language stats percentage using largest remainder

### LINKED ISSUE #22023: Language statistics percentages don't add up to 100%
### Description

This is probably just a rounding bug, but it is still a little bit frustrating when you notice it.
As you can see from the screenshot, the percentages are
COBOL 45.7%
Assembly 43.2%
Shell 5.9%
Ruby 5.3%
which is 100.1% in total.

I've made a copy of my repo here:
https://try.gitea.io/hr98/aoc22-copy-percentage-bug

### Gitea Version

Gitea version 1.17.3 built with GNU Make 4.1, go1.18.7 : bindata, sqlite, sqlite_unlock_notify

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

image

### Git Version

git version 2.31.1

### Operating System

Linux

### How are you running Gitea?

Downloaded the binary and running as systemd service on Alma Linux 9.

### Database

SQLite

### PR DESCRIPTION
**Pull request by hr-98**
Fix #22023 

I've changed how the percentages for the language statistics are rounded because they did not always add up to 100%
Now it's done with the largest remainder method, which makes sure that total is 100%

### CODE REVIEW

File: models/repo/language_stats.go [hunk 1]
```diff
@@ -43,29 +44,66 @@ func (stats LanguageStatList) LoadAttributes() {
...
+	roundByLargestRemainder(langPerc, 100)
 	return langPerc
 }
 
+// Rounds to 1 decimal point, target should be the expected sum of percs
+func roundByLargestRemainder(percs map[string]float32, target float32) {
+	// Intermediates are the floored numbers multiplied to the decimal
+	// 54.37 -> 543
+	intermediates := make(map[string]int, len(percs))
+	remainders := make(map[string]float64, len(percs))
```
Comment (zeripath) [inline models/repo/language_stats.go]: I think you can get away without these two maps by just using the percs map directly.

Comment (eeyrjmr) [reply models/repo/language_stats.go:75]: You probably could but would need to loop through twice, once to sum up the int and the 2nd to += the highest fractional

Comment (hr-98) [reply models/repo/language_stats.go]: Thanks, I can see that I can store the remainders there.
I don't see how to avoid using the intermediates map.

Comment (zeripath) [reply models/repo/language_stats.go]: A floating point number can be thought of as two numbers in one, the integer part and the remainder.

You just keep both parts around, just use the integer part as intermediates and the remainder part as the remainder.

**Review by zeripath — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Comment by lunny** [comment]
Make L-G-T-M work

**Pull request merged**